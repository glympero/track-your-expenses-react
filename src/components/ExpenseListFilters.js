import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }

    onDatesChange = ({ startDate, endDate}) => {
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({
            calendarFocused
        }));
    }

    render() {
        return (
            <div>
                <input type='text' value={this.props.filters.text} onChange={(event) => {
                    this.props.dispatch(setTextFilter(event.target.value));
                    //console.log(event.target.value);
                }} />
                <select value={this.props.filters.sortBy} onChange={(e) => {
                    const filter = e.target.value;
                    if(filter === 'date') {
                        this.props.dispatch(sortByDate());
                    }else if(filter === 'amount') {
                        this.props.dispatch(sortByAmount());
                    }
                }}>
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate} 
                    endDate={this.props.filters.endDate} 
                    onDatesChange={this.onDatesChange} 
                    focusedInput={this.state.calendarFocused} 
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
            </div>
        )
    }
    
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

export default connect(mapStateToProps)(ExpenseListFilters);

 