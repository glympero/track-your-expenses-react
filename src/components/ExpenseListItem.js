import React from 'react';

import { Link } from 'react-router-dom';

const ExpenseListItem = ({id, description, amount, createdAt}) => {
    return (
        <div>
            <h4><Link to={`/edit/${id}`}>{description}</Link></h4>
            <p>Cost: {amount}, created At: {createdAt}</p>
        </div>
    );
};

export default ExpenseListItem;