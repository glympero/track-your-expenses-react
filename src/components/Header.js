import React from 'react';
import { NavLink } from 'react-router-dom';
const Header = () => (
    <header>
      <h1>Expensify</h1>
      <div className="container">
        <NavLink to="/" className="item" activeClassName="is-active" exact={true}>Dashboard</NavLink>
        <NavLink to="/create" className="item" activeClassName="is-active">Create Expense</NavLink>
        <NavLink to="/help" className="item" activeClassName="is-active">Help Page</NavLink>
      </div>
    </header>
  );

export default Header;