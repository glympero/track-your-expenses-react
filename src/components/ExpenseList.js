import React from 'react';
import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';

import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => {
    return (
        <div>
        <h3>Expenses List</h3>
            {
                props.expenses.length === 0 ? (
                    <p>No expenses Found</p>
                ) : (
                    
                    props.expenses.map((expense) => {
                        return (
                            // <ExpenseListItem 
                            //     key={expense.id} 
                            //     description={expense.description}
                            //     amount={expense.amount}
                            //     createdAt={expense.createdAt}
                            // />
                            <ExpenseListItem key={expense.id}  {...expense} />
                        );
                    })
                )
            }
            {}
        </div>        
    );
};

// const ConnectedExpenseList = connect((state) => {
//     console.log(state.expenses)
//     return {
//         expenses: state.expenses
//     };
// })(ExpenseList);

// export default ConnectedExpenseList;

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};

export default connect(mapStateToProps)(ExpenseList);
