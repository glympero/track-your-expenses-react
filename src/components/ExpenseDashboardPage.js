import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = (props) => {
    return (
      <div>
      <h1>This is my dashboard component</h1>
      <ExpenseListFilters />
      <ExpenseList />
    </div>
    );
  };

export default ExpenseDashboardPage;