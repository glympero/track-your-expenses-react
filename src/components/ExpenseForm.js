import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import ExpenseDashboardPage from './ExpenseDashboardPage';

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({
            description
        }));
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({
            note
        }));
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        /*eslint no-control-regex: "error"*/
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) { //If there is no amount or amount is matched
            this.setState(() => ({
                amount
            }));
        }        
    }

    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState(() => ({
                createdAt
            }));
        }
    }

    onFocusChange = ( { focused }) => {
        this.setState(() => ({
            calendarFocused: focused
        }));
    }

    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.description || !this.state.amount) {
            this.setState(() => ({
                error: 'Please provide description and amount'
            }));
        }else {
            this.setState(() => ({
                error: ''
            }));
            const description = this.state.description;
            const note = this.state.note;
            const amount = parseFloat(this.state.amount, 10) * 100;
            const createdAt = this.state.createdAt.valueOf();
            this.props.onSubmit({ //this is the function passed from addExpense component
                description,
                note,
                amount,
                createdAt
            });
        }
    }

    render() {
        return (
            <div>
                <h2>Expense Form</h2>
                { !!this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input type='text' value={this.state.description} onChange={this.onDescriptionChange} placeholder='Description' autoFocus/>
                    <input type='text' value={this.state.amount} onChange={this.onAmountChange} placeholder='amount' />
                    <textarea value={this.state.note} onChange={this.onNoteChange} placeholder='Add a note for your expense (optional)'>
                    </textarea>
                    <SingleDatePicker
                        date={this.state.createdAt} 
                        onDateChange={this.onDateChange} 
                        focused={this.state.calendarFocused} 
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1} //Show only one month
                        isOutsideRange={(day) => false} //Select days before current day
                    />
                     <button 
                        //disabled={!this.state.description || !this.state.amount}
                    >Submit</button>
                </form>
            </div>
        );
    }
}

export default ExpenseForm;