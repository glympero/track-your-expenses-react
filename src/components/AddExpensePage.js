import React from 'react';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';
import { connect } from 'react-redux';

export class AddExpensePage extends React.Component {

  onSubmit = (expense) => {
    //this will be called from ExpenseForm and will provide an expense object
    //props.dispatch(addExpense(expense));
    this.props.addExpense(expense)
    this.props.history.push('/');
  }

  render(){
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    )
  }
}

  //maps dispatch actions to props
  //it has access to dispatch
  //we return an object where we define various props
const mapDispatchToProps = (dispatch) => {
  return {
    addExpense: (expense) => {
      dispatch(addExpense(expense));
    }
  }
}

export default connect(null, mapDispatchToProps)(AddExpensePage);