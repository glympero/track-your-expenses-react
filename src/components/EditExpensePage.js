import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { editExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {

  onRemove = () => {
    //props.dispatch(removeExpense({ id }));
    this.props.startRemoveExpense( {id: this.props.expense.id} );
    this.props.history.push('/');
  }

  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
      <h1>This is my edit expense component</h1>
      <p>Editing item with the id of {this.props.match.params.id}</p>
      <ExpenseForm
        onSubmit={this.onSubmit}
        expense={this.props.expense}
      />
      <button onClick={this.onRemove}>Remove</button>
    </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  let expense = state.expenses.find( e => e.id === props.match.params.id);
  return {
    expense
  }
};

const mapDispatchToProps = (dispatch, props) => {
  return {
    startRemoveExpense: (data) => {
      dispatch(startRemoveExpense(data));
    },
    editExpense: (id, expense) => {
      dispatch(editExpense(id, expense));
    }
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);