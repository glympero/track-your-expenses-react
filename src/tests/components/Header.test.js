// import ReactShallowRenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme';
import React from 'react';

import { Header } from '../../components/Header';

// test('should render header correctly', () => {
//     const renderer = new ReactShallowRenderer();

//     renderer.render(<Header />);
//     expect(renderer.getRenderOutput()).toMatchSnapshot();
// });


test('should render header correctly', () => {
    const wrapper = shallow(<Header startLogout={() => {}} />);

    expect(wrapper.find('h1').length).toBe(1);
    expect(wrapper.find('h1').text()).toBe('Expensify');

    expect(wrapper).toMatchSnapshot();
});

test('should call startLogout', () => {
    const startLogout = jest.fn();
    const wrapper = shallow(<Header startLogout={startLogout} />);
    wrapper.find('button').prop('onClick')();
    //history is our spy above and we check if it has been called with the correct data passed in.
    expect(startLogout).toHaveBeenCalled();
});

//should call startLogout on button click