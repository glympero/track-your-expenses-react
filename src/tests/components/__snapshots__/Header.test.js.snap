// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render header correctly 1`] = `
<header>
  <h1>
    Expensify
  </h1>
  <div
    className="container"
  >
    <NavLink
      activeClassName="is-active"
      ariaCurrent="true"
      className="item"
      exact={true}
      to="/"
    >
      Dashboard
    </NavLink>
    <NavLink
      activeClassName="is-active"
      ariaCurrent="true"
      className="item"
      to="/create"
    >
      Create Expense
    </NavLink>
    <NavLink
      activeClassName="is-active"
      ariaCurrent="true"
      className="item"
      to="/help"
    >
      Help Page
    </NavLink>
  </div>
</header>
`;
